keps/sig-storage/1495-volume-populators/README.md

```diff
@@ -105,21 +91,48 @@ flexibility will allow us to experiment with approaches for data populators.
 - Enable users to create pre-populated volumes in a manner consistent with current practice
 - Enable developers to innovate with new an interesting data sources
 - Avoid changing existing workflows and breaking existing tools with new behavior
+- Ensure that users continue to get accurate error messages and feedback about volume
+  creation progress, whether using existing methods of volume creation or specifying a
+  data source associated with a volume populator.
 
 ### Non-Goals
 
 - This proposal DOES NOT recommend any specific approach for data populators. The specific
   design of how a data populator should work will be handled in a separate KEP.
-- We do not propose any new object types. Specifically, no special "populator"
-  object type. Populators should function based on their own object types.
 
 ## Proposal
 
-The validation for `DataSource` field should be removed entirely, in the fullness of time.
-
-Short term, we should add a new alpha feature gate, which allows arbitrary objects to be
-specified for the `DataSource` field, with the intention to eventually make this behavior
-the standard. 
+Validation for the `DataSource` field should be moved from the core to a validation
+webhook. Because the existing validation only allow snapshots and volumes as data sources,
+a feature gate is added to relax that validation, allowing the existing object types plus
+any `Kind` of CR. Each volume populator should be able to register one of more kinds of
+objects that it understands.
+
+In order to ensure that dead-on-arrival PVCs (ones that will never be bound) can't be
+created, the webhook needs to deny creation of PVCs with a `DataSource` `Kind` that it
+doesn't recognize. This ensures that PVCs which do get created will have some controller
```

Comment:
But the registration of the (new)populator or adding a new datasource on existing populator can happen laetr too, in that case, stopping it upfront based on the webhook is against the kube norm ie dsw. Isnt it ? Did I fail to interpret this ?

Reply:
You're right that this is a slightly different pattern. But the existing behavior (silently dropping the datasource) is also kind of against the normal pattern.

Given that the enforcement mechanism for telling users their data source is invalid will be a validating webhook rather than a full-blown controller, denying the object creation is really the only action we can take.

If there are use cases for adding a new populator to a running cluster (as opposed to setting it up at cluster creation time) then users who wish to consume that populator will need to directly check for the existence of the populator CR and wait until it's present before creating PVCs that use that data source (or just loop on failure).